### PR TITLE
fix LS stakers

### PR DESCRIPTION
**Pull request by xsteadybcgo**
## Context

* Logic error correction

## Were any new external packages added?

* List external packages added and why they were added

### CODE REVIEW

File: packages/app/src/controller/staking/metadata.ts [hunk 1]
```diff
@@ -11,35 +11,64 @@ const orderByBlockHeightDesc: FindOneOptions = {
...
+    ...res
   })
   return next()
 }
 
 const getListWithTotalStakers = async (list: any[], network: string): Promise<any[]> => {
   let lastValidStakeCounts: number = 0
 
+  // Find the first block height with stake action
+  const latestBlockHeight = list[0].blockHeight
```
Comment (zhihuizhangjaden) [inline packages/app/src/controller/staking/metadata.ts:50]: Confused about this

File: packages/app/src/controller/staking/metadata.ts [hunk 2]
```diff
@@ -11,35 +11,64 @@ const orderByBlockHeightDesc: FindOneOptions = {
   }
 }
 
+const getDataWithTotalStakers = async (data: any, network: string): Promise<any> => {
```
Comment (zhihuizhangjaden) [inline packages/app/src/controller/staking/metadata.ts:14]: Cloud we change this method to be immutable way?

File: packages/app/src/controller/staking/metadata.ts (hunk 1 above)
Comment (xsteadybcgo) [reply packages/app/src/controller/staking/metadata.ts:50]: > Confused about this

TLDR: totalStaker comes from a redis hash table, not every block height has data, its maximum block height depends on last `Staked` event, but the maximum block height of the metadata may be larger than the block height of the hash table, in order to ensure that all metadata have totalStaker, we take the highest block of the metadata and traverse to the first block with totalStaker in the hash table to fill in the subsequent data.

This part of the logic is really complex due to the data separation (combine stakers with metadata data but their block heights don't match) and in the next version to refactor this part of the data, which will be bundled together to reduce the complexity of the logic :)

File: packages/app/src/controller/staking/metadata.ts (hunk 2 above)
Comment (xsteadybcgo) [reply packages/app/src/controller/staking/metadata.ts:14]: From
 ```
  data.totalStakers = Number(totalStakers)
  return data

```
to 

`return { ...data, totalStakers: Number(totalStakers) }` ？

Not really much difference here? I think, the intent here is to do overrides, due to duplicate data sources. Wait for refactoring to solve these problems and this functions no longer needed.

File: packages/app/src/controller/staking/metadata.ts [hunk 3]
```diff
@@ -11,35 +11,64 @@ const orderByBlockHeightDesc: FindOneOptions = {
...
+  // Find the first block height with stake action
+  for (let i = data.blockHeight; i >= 0; i--) {
+    const totalStakers = await cacheRd.hget(
+      KEYS.Cache.hStakingStakeBlockHeight(network),
+      i.toString()
+    )
+    if (!!totalStakers) {
```
Comment (BruceChar) [inline packages/app/src/controller/staking/metadata.ts:21]: what if totalStakers = '0'?

Comment (BruceChar) [reply packages/app/src/controller/staking/metadata.ts:21]: does the key value has default value?

Comment (BruceChar) [reply packages/app/src/controller/staking/metadata.ts:21]: it is better to define a bool var like `const hasStaker: boolean = totalStakers !== undefined && totalStakers !== '0' `

File: packages/app/src/controller/staking/metadata.ts [hunk 4]
```diff
@@ -11,35 +11,64 @@ const orderByBlockHeightDesc: FindOneOptions = {
...
   const data = await ParaStakingMetadata.findOne(orderByBlockHeightDesc)
+  const res = await getDataWithTotalStakers(data, 'para')
   ctx.body = Resp.Ok({
-    ...data
+    ...res
   })
   return next()
 }
 
 const getListWithTotalStakers = async (list: any[], network: string): Promise<any[]> => {
```
Comment (BruceChar) [inline packages/app/src/controller/staking/metadata.ts:46]: what is the list type? can we replace any to explict type?

File: packages/app/src/controller/staking/metadata.ts [hunk 5]
```diff
@@ -11,35 +11,64 @@ const orderByBlockHeightDesc: FindOneOptions = {
...
   let lastValidStakeCounts: number = 0
 
+  // Find the first block height with stake action
+  const latestBlockHeight = list[0].blockHeight
+  for (let i = latestBlockHeight; i >= 0; i--) {
+    const validStakers = await cacheRd.hget(
```
Comment (BruceChar) [inline packages/app/src/controller/staking/metadata.ts:52]: same above！here you named the var `validStakers` like `isValidStakers` but it's a number string actually and used for condition assert. a little confused, validStakersNumber or valideStakersCount maybe make sense

**Comment by BruceChar** [comment]
maybe we need extract out this logic to polling

File: packages/app/src/controller/staking/metadata.ts (hunk 3 above)
Comment (xsteadybcgo) [reply packages/app/src/controller/staking/metadata.ts:21]: Not necessary, If the block does not exist, `null` will be returned and a ` '0' ` should not be treated as invalid data, although it will not appear to be `'0'`

File: packages/app/src/controller/staking/metadata.ts (hunk 4 above)
Comment (xsteadybcgo) [reply packages/app/src/controller/staking/metadata.ts:46]: right, fix it

File: packages/app/src/controller/staking/metadata.ts (hunk 5 above)
Comment (xsteadybcgo) [reply packages/app/src/controller/staking/metadata.ts:52]: As stated above ^

**Comment by xsteadybcgo** [comment]
> maybe we need extract out this logic to polling

Absolutely yes, I'm also suffering from it.

**Review by zhihuizhangjaden — APPROVED** [review]
